<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Mohammed Omar Elsiddieg Abdallah" />

<meta name="date" content="2016-06-19" />

<title>Simplifying Genomic Annotations in R</title>


<style type="text/css">code{white-space: pre;}</style>
<link href="data:text/css;charset=utf-8,pre%20%2Eoperator%2C%0Apre%20%2Eparen%20%7B%0Acolor%3A%20rgb%28104%2C%20118%2C%20135%29%0A%7D%0Apre%20%2Eliteral%20%7B%0Acolor%3A%20%23990073%0A%7D%0Apre%20%2Enumber%20%7B%0Acolor%3A%20%23099%3B%0A%7D%0Apre%20%2Ecomment%20%7B%0Acolor%3A%20%23998%3B%0Afont%2Dstyle%3A%20italic%0A%7D%0Apre%20%2Ekeyword%20%7B%0Acolor%3A%20%23900%3B%0Afont%2Dweight%3A%20bold%0A%7D%0Apre%20%2Eidentifier%20%7B%0Acolor%3A%20rgb%280%2C%200%2C%200%29%3B%0A%7D%0Apre%20%2Estring%20%7B%0Acolor%3A%20%23d14%3B%0A%7D%0A" rel="stylesheet" type="text/css" />
<script src="data:application/x-javascript;base64,
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();

"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>


<link href="data:text/css;charset=utf-8,body%2C%20td%20%7B%0Afont%2Dfamily%3A%20sans%2Dserif%3B%0Abackground%2Dcolor%3A%20white%3B%0Afont%2Dsize%3A%2013px%3B%0A%7D%0Abody%20%7B%0Amax%2Dwidth%3A%20800px%3B%0Amargin%3A%200%20auto%3B%0Apadding%3A%201em%201em%202em%3B%0Aline%2Dheight%3A%2020px%3B%0A%7D%0A%0Adiv%23TOC%20li%20%7B%0Alist%2Dstyle%3Anone%3B%0Abackground%2Dimage%3Anone%3B%0Abackground%2Drepeat%3Anone%3B%0Abackground%2Dposition%3A0%3B%0A%7D%0A%0Ap%2C%20pre%20%7B%20margin%3A%200em%200em%201em%3B%0A%7D%0A%0Aimg%2C%20table%20%7B%0Amargin%3A%200em%20auto%201em%3B%0A%7D%0Ap%20%7B%0Atext%2Dalign%3A%20justify%3B%0A%7D%0Att%2C%20code%2C%20pre%20%7B%0Afont%2Dfamily%3A%20%27DejaVu%20Sans%20Mono%27%2C%20%27Droid%20Sans%20Mono%27%2C%20%27Lucida%20Console%27%2C%20Consolas%2C%20Monaco%2C%20monospace%3B%0A%7D%0Ah1%2C%20h2%2C%20h3%2C%20h4%2C%20h5%2C%20h6%20%7B%20font%2Dfamily%3A%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Amargin%3A%201%2E2em%200em%200%2E6em%200em%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Ah1%2Etitle%20%7B%0Afont%2Dsize%3A%20250%25%3B%0Afont%2Dweight%3A%20normal%3B%0Acolor%3A%20%2387b13f%3B%0Aline%2Dheight%3A%201%2E1em%3B%0Amargin%2Dtop%3A%200px%3B%0Aborder%2Dbottom%3A%200px%3B%0A%7D%0Ah1%20%7B%0Afont%2Dsize%3A%20160%25%3B%0Afont%2Dweight%3A%20normal%3B%0Aline%2Dheight%3A%201%2E4em%3B%0Aborder%2Dbottom%3A%201px%20%231a81c2%20solid%3B%0A%7D%0Ah2%20%7B%0Afont%2Dsize%3A%20130%25%3B%20%7D%0Ah1%2C%20h2%2C%20h3%20%7B%0Acolor%3A%20%231a81c2%3B%0A%7D%0Ah3%2C%20h4%2C%20h5%2C%20h6%20%7B%0Afont%2Dsize%3A115%25%3B%0A%7D%20%0A%0Aa%20%7B%20color%3A%20%231a81c2%3B%20%7D%0Aa%3Aactive%20%7B%20outline%3A%20none%3B%20%7D%0Aa%3Avisited%20%7B%20color%3A%20%231a81c2%3B%20%7D%0Aa%3Ahover%20%7B%20color%3A%20%234c94c2%3B%20%7D%0Apre%2C%20img%20%7B%0Amax%2Dwidth%3A%20100%25%3B%0Adisplay%3A%20block%3B%0A%7D%0Apre%20%7B%0Aborder%3A%200px%20none%3B%0Abackground%2Dcolor%3A%20%23F8F8F8%3B%0Awhite%2Dspace%3A%20pre%3B%0Aoverflow%2Dx%3A%20auto%3B%0A%7D%0Apre%20code%20%7B%0Aborder%3A%201px%20%23aaa%20dashed%3B%0Abackground%2Dcolor%3A%20white%3B%0Adisplay%3A%20block%3B%0Apadding%3A%201em%3B%20color%3A%20%23111%3B%0Aoverflow%2Dx%3A%20inherit%3B%0A%7D%0A%0Apre%20code%5Bclass%5D%20%7B%0Abackground%2Dcolor%3A%20inherit%3B%0A%7D%0A%0Apre%5Bclass%5D%20code%20%7B%0Abackground%2Dcolor%3A%20inherit%3B%0A%7D%0A%0Acode%20%7B%20background%2Dcolor%3A%20transparent%3B%0Acolor%3A%20%2387b13f%3B%0Afont%2Dsize%3A%2092%25%3B%0A%7D%0A%0Atable%2C%20td%2C%20th%20%7B%0Aborder%3A%20none%3B%0Apadding%3A%200%200%2E5em%3B%0A%7D%0A%0Atbody%20tr%3Anth%2Dchild%28odd%29%20td%20%7B%0Abackground%2Dcolor%3A%20%23F8F8F8%3B%0A%7D%0Ablockquote%20%7B%0Acolor%3A%23666666%3B%0Amargin%3A0%3B%0Apadding%2Dleft%3A%201em%3B%0Aborder%2Dleft%3A%200%2E5em%20%23EEE%20solid%3B%0A%7D%0Ahr%20%7B%0Aheight%3A%200px%3B%0Aborder%2Dbottom%3A%20none%3B%0Aborder%2Dtop%2Dwidth%3A%20thin%3B%0Aborder%2Dtop%2Dstyle%3A%20dotted%3B%0Aborder%2Dtop%2Dcolor%3A%20%23999999%3B%0A%7D%0Aspan%2Eheader%2Dsection%2Dnumber%20%7B%0Apadding%2Dright%3A%201em%3B%0A%7D%0Aspan%2Etoc%2Dsection%2Dnumber%3A%3Aafter%20%7B%0Acontent%3A%20%22%20%22%3B%0Awhite%2Dspace%3A%20pre%3B%0A%7D%0A%40media%20print%20%7B%0A%2A%20%7B%0Abackground%3A%20transparent%20%21important%3B%0Acolor%3A%20black%20%21important%3B%0Afilter%3Anone%20%21important%3B%0A%2Dms%2Dfilter%3A%20none%20%21important%3B%0A%7D%0Abody%20%7B%0Afont%2Dsize%3A12pt%3B%0Amax%2Dwidth%3A100%25%3B%0A%7D%0Aa%2C%20a%3Avisited%20%7B%0Atext%2Ddecoration%3A%20underline%3B%0A%7D%0Ahr%20%7B%0Avisibility%3A%20hidden%3B%0Apage%2Dbreak%2Dbefore%3A%20always%3B%0A%7D%0Apre%2C%20blockquote%20%7B%0Apadding%2Dright%3A%201em%3B%0Apage%2Dbreak%2Dinside%3A%20avoid%3B%0A%7D%0Atr%2C%20img%20%7B%0Apage%2Dbreak%2Dinside%3A%20avoid%3B%0A%7D%0Aimg%20%7B%0Amax%2Dwidth%3A%20100%25%20%21important%3B%0A%7D%0A%40page%20%3Aleft%20%7B%0Amargin%3A%2015mm%2020mm%2015mm%2010mm%3B%0A%7D%0A%40page%20%3Aright%20%7B%0Amargin%3A%2015mm%2010mm%2015mm%2020mm%3B%0A%7D%0Ap%2C%20h2%2C%20h3%20%7B%0Aorphans%3A%203%3B%20widows%3A%203%3B%0A%7D%0Ah2%2C%20h3%20%7B%0Apage%2Dbreak%2Dafter%3A%20avoid%3B%0A%7D%0A%7D%0A" rel="stylesheet" type="text/css" />

<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if(links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>

</head>

<body>


<div id="header">
<h1 class="title">Simplifying Genomic Annotations in R</h1>
<h4 class="author"><em>Mohammed Omar Elsiddieg Abdallah</em></h4>
<h4 class="date"><em>19 June 2016</em></h4>
</div>
<h4 class="package">Package version: <span style="font-weight: normal">BiocStyle 2.1.7</span></h4>

<h1>Contents</h1>
<div id="TOC">
<ul>
<li><a href="#introduction-to-cellbaser"><span class="toc-section-number">1</span> Introduction to CellbaseR</a></li>
<li><a href="#cellbaser-classes-and-methods"><span class="toc-section-number">2</span> CellbaseR Classes and Methods</a><ul>
<li><a href="#the-cellbaser-class"><span class="toc-section-number">2.1</span> The CellbaseR class</a></li>
<li><a href="#the-cellbaser-methods"><span class="toc-section-number">2.2</span> The CellbaseR Methods</a><ul>
<li><a href="#cbgeneclient"><span class="toc-section-number">2.2.1</span> cbGeneClient</a></li>
<li><a href="#cbregionclient"><span class="toc-section-number">2.2.2</span> cbRegionClient</a></li>
<li><a href="#cbvraiantclient"><span class="toc-section-number">2.2.3</span> cbVraiantClient</a></li>
<li><a href="#cbclinical"><span class="toc-section-number">2.2.4</span> cbClinical</a></li>
</ul></li>
</ul></li>
<li><a href="#cellbaser-utilities"><span class="toc-section-number">3</span> CellbaseR utilities</a><ul>
<li><a href="#creategenemodel"><span class="toc-section-number">3.1</span> CreateGeneModel</a></li>
<li><a href="#cbannotatevcf"><span class="toc-section-number">3.2</span> cbAnnotateVcf</a></li>
</ul></li>
</ul>
</div>

<div id="introduction-to-cellbaser" class="section level1">
<h1><span class="header-section-number">1</span> Introduction to CellbaseR</h1>
<p>This R package makes use of the exhaustive RESTful Web service API that has been implemented for the Cellabase database. It enable researchers to query and obtain a wealth of biological information from a single database saving a lot of time. Another benefit is that researchers can easily make queries about different biological topics and link all this information together as all information is integrated.</p>
</div>
<div id="cellbaser-classes-and-methods" class="section level1">
<h1><span class="header-section-number">2</span> CellbaseR Classes and Methods</h1>
<div id="the-cellbaser-class" class="section level2">
<h2><span class="header-section-number">2.1</span> The CellbaseR class</h2>
<p>This is an S4 class that holds the basic configuration for connecting to the Cellbase web services. Let us start by loading the library.</p>
<pre class="r"><code># to get the default CellbaseR object (human data, from genome GRCh37)
library(cellbaseR)
# A default cellbaseR object is created as follows
cb &lt;- CellBaseR()
# to change the default species from human to mouse for example
mm &lt;-CellBaseR(species = &quot;mmsculus&quot;) </code></pre>
<p>You can see the avaiable species and their corresonding data by running the cbSpecies method passing the CellbBseR object you just created</p>
<pre class="r"><code>library(cellbaseR)
cb &lt;- CellBaseR()
sp &lt;- cbSpeciesClient(object = cb)
# This will give you a CellbaseResult object
# to get the dataframe of all available species
sp &lt;- cbData(sp)
str(sp,1)</code></pre>
<pre><code>## 'data.frame':    48 obs. of  4 variables:
##  $ id            : chr  &quot;hsapiens&quot; &quot;mmusculus&quot; &quot;drerio&quot; &quot;rnorvegicus&quot; ...
##  $ scientificName: chr  &quot;Homo sapiens&quot; &quot;Mus musculus&quot; &quot;Danio rerio&quot; &quot;Rattus norvegicus&quot; ...
##  $ assemblies    :List of 48
##  $ data          :List of 48</code></pre>
</div>
<div id="the-cellbaser-methods" class="section level2">
<h2><span class="header-section-number">2.2</span> The CellbaseR Methods</h2>
<p>In addition to the cbSpecies you have seen already, The cellbaseR package provide many methods to query the cellbase webservices, they include:</p>
<ul>
<li>cbGeneClient<br />
</li>
<li>cbSnpClient<br />
</li>
<li>cbProteinClient<br />
</li>
<li>cbTranscriptClient</li>
<li>cbRegionClient</li>
<li>cbVariantClient<br />
</li>
<li>cbXrefClient<br />
</li>
<li>cbClinicalClient</li>
</ul>
<p>In all cases the user is expected to provide the ids for the query and the resource to be queried, in addition to the CellbaseQuery object they created. For example, a query through the cbGene will look something like this</p>
<div id="cbgeneclient" class="section level3">
<h3><span class="header-section-number">2.2.1</span> cbGeneClient</h3>
<pre class="r"><code>library(cellbaseR)
cb &lt;- CellBaseR()
genes &lt;- c(&quot;TP73&quot;,&quot;TET1&quot;)

res &lt;- cbGeneClient(object = cb, ids = genes, resource = &quot;info&quot;)
# to get the resulting data.frame run cellbaseData()
res &lt;- cbData(object = res)
str(res,2)</code></pre>
<pre><code>## 'data.frame':    2 obs. of  12 variables:
##  $ id         : chr  &quot;ENSG00000078900&quot; &quot;ENSG00000138336&quot;
##  $ name       : chr  &quot;TP73&quot; &quot;TET1&quot;
##  $ biotype    : chr  &quot;protein_coding&quot; &quot;protein_coding&quot;
##  $ status     : chr  &quot;KNOWN&quot; &quot;KNOWN&quot;
##  $ chromosome : chr  &quot;1&quot; &quot;10&quot;
##  $ start      : int  3569084 70320413
##  $ end        : int  3652765 70454239
##  $ strand     : chr  &quot;+&quot; &quot;+&quot;
##  $ source     : chr  &quot;Ensembl&quot; &quot;Ensembl&quot;
##  $ description: chr  &quot;tumor protein p73 [Source:HGNC Symbol;Acc:HGNC:12003]&quot; &quot;tet methylcytosine dioxygenase 1 [Source:HGNC Symbol;Acc:HGNC:29484]&quot;
##  $ transcripts:List of 2
##   ..$ :'data.frame': 14 obs. of  20 variables:
##   ..$ :'data.frame': 1 obs. of  20 variables:
##  $ annotation :'data.frame': 2 obs. of  2 variables:
##   ..$ expression:List of 2
##   ..$ diseases  :List of 2</code></pre>
<pre class="r"><code># as you can see the res dataframe also contains a transcriots column 
# which is in fact a list column of nested dataframes, to get the
# trasnscripts data.frame for first gene
TET1_transcripts &lt;- res$transcripts[[1]]
str(TET1_transcripts,1)</code></pre>
<pre><code>## 'data.frame':    14 obs. of  20 variables:
##  $ id                : chr  &quot;ENST00000378295&quot; &quot;ENST00000604074&quot; &quot;ENST00000354437&quot; &quot;ENST00000357733&quot; ...
##  $ name              : chr  &quot;TP73-001&quot; &quot;TP73-006&quot; &quot;TP73-002&quot; &quot;TP73-202&quot; ...
##  $ biotype           : chr  &quot;protein_coding&quot; &quot;protein_coding&quot; &quot;protein_coding&quot; &quot;protein_coding&quot; ...
##  $ status            : chr  &quot;KNOWN&quot; &quot;KNOWN&quot; &quot;KNOWN&quot; &quot;KNOWN&quot; ...
##  $ chromosome        : chr  &quot;1&quot; &quot;1&quot; &quot;1&quot; &quot;1&quot; ...
##  $ start             : int  3569084 3569084 3569129 3569129 3569129 3598930 3598930 3607236 3607236 3607236 ...
##  $ end               : int  3652765 3652765 3649856 3650467 3650467 3649643 3649643 3649856 3649856 3652765 ...
##  $ strand            : chr  &quot;+&quot; &quot;+&quot; &quot;+&quot; &quot;+&quot; ...
##  $ genomicCodingStart: int  3598930 3598930 3598930 3598930 3598930 3598930 3598930 3607470 3607470 3607470 ...
##  $ genomicCodingEnd  : int  3649643 3649326 3649326 3649643 3649643 3649643 3649643 3649326 3648119 3649643 ...
##  $ cdnaCodingStart   : int  156 156 111 111 111 1 1 235 235 235 ...
##  $ cdnaCodingEnd     : int  2066 1367 1610 1778 1733 1668 1623 1587 1515 1998 ...
##  $ cdsLength         : int  1910 1211 1499 1667 1622 1667 1622 1352 1280 1763 ...
##  $ proteinID         : chr  &quot;ENSP00000367545&quot; &quot;ENSP00000475143&quot; &quot;ENSP00000346423&quot; &quot;ENSP00000350366&quot; ...
##  $ proteinSequence   : chr  &quot;MAQSTATSPDGGTTFEHLWSSLEPDSTYFDLPQSSRGNNEVVGGTDSSMDVFHLEGMTTSVMAQFNLLSSTMDQMSSRAASASPYTPEHAASVPTHSPYAQPSSTFDTMSPAPVIPSNTDYPGPHHF&quot;| __truncated__ &quot;MAQSTATSPDGGTTFEHLWSSLEPDSTYFDLPQSSRGNNEVVGGTDSSMDVFHLEGMTTSVMAQFNLLSSTMDQMSSRAASASPYTPEHAASVPTHSPYAQPSSTFDTMSPAPVIPSNTDYPGPHHF&quot;| __truncated__ &quot;MAQSTATSPDGGTTFEHLWSSLEPDSTYFDLPQSSRGNNEVVGGTDSSMDVFHLEGMTTSVMAQFNLLSSTMDQMSSRAASASPYTPEHAASVPTHSPYAQPSSTFDTMSPAPVIPSNTDYPGPHHF&quot;| __truncated__ &quot;MAQSTATSPDGGTTFEHLWSSLEPDSTYFDLPQSSRGNNEVVGGTDSSMDVFHLEGMTTSVMAQFNLLSSTMDQMSSRAASASPYTPEHAASVPTHSPYAQPSSTFDTMSPAPVIPSNTDYPGPHHF&quot;| __truncated__ ...
##  $ cDnaSequence      : chr  &quot;TGCCTCCCCGCCCGCGCACCCGCCCGGAGGCTCGCGCGCCCGCGAAGGGGACGCAGCGAAACCGGGGCCCGCGCCAGGCCAGCCGGGACGGACGCCGATGCCCGGGGCTGCGACGGCTGCAGAGCGA&quot;| __truncated__ &quot;TGCCTCCCCGCCCGCGCACCCGCCCGGAGGCTCGCGCGCCCGCGAAGGGGACGCAGCGAAACCGGGGCCCGCGCCAGGCCAGCCGGGACGGACGCCGATGCCCGGGGCTGCGACGGCTGCAGAGCGA&quot;| __truncated__ &quot;AGGGGACGCAGCGAAACCGGGGCCCGCGCCAGGCCAGCCGGGACGGACGCCGATGCCCGGGGCTGCGACGGCTGCAGAGCGAGCTGCCCTCGGAGGCCGGCGTGGGGAAGATGGCCCAGTCCACCGC&quot;| __truncated__ &quot;AGGGGACGCAGCGAAACCGGGGCCCGCGCCAGGCCAGCCGGGACGGACGCCGATGCCCGGGGCTGCGACGGCTGCAGAGCGAGCTGCCCTCGGAGGCCGGCGTGGGGAAGATGGCCCAGTCCACCGC&quot;| __truncated__ ...
##  $ xrefs             :List of 14
##  $ tfbs              :List of 14
##  $ exons             :List of 14
##  $ annotationFlags   :List of 14</code></pre>
</div>
<div id="cbregionclient" class="section level3">
<h3><span class="header-section-number">2.2.2</span> cbRegionClient</h3>
<pre class="r"><code># making a query through cbRegion to get all the clinically relevant variants 
# in a specific region
library(cellbaseR)
cb &lt;- CellBaseR()
res &lt;- cbRegionClient(object=cb,ids=&quot;17:1000000-1100000&quot;,
resource=&quot;clinical&quot;)
# to get all conservation data in this region
res &lt;- cbRegionClient(object=cb,ids=&quot;17:1000000-1100000&quot;,
resource=&quot;conservation&quot;)
#likewise to get all the regulatory data for the same region
res &lt;- cbRegionClient(object=cb,ids=&quot;17:1000000-1100000&quot;, resource=&quot;regulatory&quot;)
res &lt;- cbData(res)
str(res,1)</code></pre>
<pre><code>## 'data.frame':    959 obs. of  13 variables:
##  $ chromosome  : chr  &quot;17&quot; &quot;17&quot; &quot;17&quot; &quot;17&quot; ...
##  $ source      : chr  &quot;ccat_histone&quot; &quot;ccat_histone&quot; &quot;SWEmbl_R0005_IDR&quot; &quot;ccat_histone&quot; ...
##  $ featureType : chr  &quot;H3K36_trimethylation_site&quot; &quot;H3K36_trimethylation_site&quot; &quot;H3K27_acylation_site&quot; &quot;H3K36_trimethylation_site&quot; ...
##  $ start       : int  1000780 1000860 1000735 1000780 1000820 999880 1000260 1000900 1001500 1001840 ...
##  $ end         : int  1001300 1001120 1002110 1002200 1001340 1000680 1000500 1001200 1002240 1002160 ...
##  $ score       : chr  &quot;.&quot; &quot;.&quot; &quot;.&quot; &quot;.&quot; ...
##  $ strand      : chr  &quot;.&quot; &quot;.&quot; &quot;.&quot; &quot;.&quot; ...
##  $ frame       : chr  &quot;.&quot; &quot;.&quot; &quot;.&quot; &quot;.&quot; ...
##  $ name        : chr  &quot;H3K36me3&quot; &quot;H3K36me3&quot; &quot;H3K27ac&quot; &quot;H3K36me3&quot; ...
##  $ featureClass: chr  &quot;Histone&quot; &quot;Histone&quot; &quot;Histone&quot; &quot;Histone&quot; ...
##  $ alias       : chr  &quot;HSMM_H3K36me3_ENCODE_Broad_ccat_histone&quot; &quot;HeLa-S3_H3K36me3_ENCODE_Broad_ccat_histone&quot; &quot;Monocytes-CD14+_H3K27ac_ENCODE_Broad_SWEmbl_R0005_IDR&quot; &quot;A549_H3K36me3_ENCODE_Broad_ccat_histone&quot; ...
##  $ cellTypes   :List of 959
##   .. [list output truncated]
##  $ matrix      : chr  NA NA NA NA ...</code></pre>
</div>
<div id="cbvraiantclient" class="section level3">
<h3><span class="header-section-number">2.2.3</span> cbVraiantClient</h3>
<p>Getting annotation for a specific variant</p>
<pre class="r"><code>library(cellbaseR)
cb &lt;- CellBaseR()
res2 &lt;- cbVariantClient(object=cb, ids=&quot;1:169549811:A:G&quot;, resource=&quot;annotation&quot;)
# to get the data 
res2 &lt;- cbData(res2)
str(res2, 1)</code></pre>
<p>A very powerfull feature of cellbaseR is ability to fetch a wealth of clinical data, as well as abiliy to fiter these data by gene, phenotype, rs, etc…</p>
</div>
<div id="cbclinical" class="section level3">
<h3><span class="header-section-number">2.2.4</span> cbClinical</h3>
<pre class="r"><code>library(cellbaseR)
cb &lt;- CellBaseR()
# First we have to specify aour filters, we do that by creating an object of
# class CellbaseParam
cbparam &lt;- CellBaseParam(gene=&quot;TET1&quot;, genome=&quot;GRCh38&quot;)
cbparam</code></pre>
<pre><code>## An object of class CellBaseParam
## use this object to control what results are returned from the
##     CellBaseR methods</code></pre>
<pre class="r"><code># Note that cbClinical does NOT require any Ids to be passed, only the filters
# and of course the CellbaseQuery object
res &lt;- cbClinicalClient(object=cb,filters=cbparam)
res</code></pre>
<pre><code>## An object of class CellBaseResponse
##  containing 12 rows and 35 columns.
##  to get the annotated dataframe use cbData()</code></pre>
<pre class="r"><code>res &lt;- cbData(res)
str(res,1)</code></pre>
<pre><code>## 'data.frame':    12 obs. of  35 variables:
##  $ chromosome                  : chr  &quot;10&quot; &quot;10&quot; &quot;10&quot; &quot;10&quot; ...
##  $ start                       : int  70446203 70446178 70451227 70333081 70406255 70450816 70406413 70426853 70450697 70406107 ...
##  $ end                         : int  70446203 70446178 70451227 70333081 70406255 70450816 70406413 70426853 70450697 70406107 ...
##  $ reference                   : chr  &quot;G&quot; &quot;T&quot; &quot;G&quot; &quot;T&quot; ...
##  $ alternate                   : chr  &quot;A&quot; &quot;A&quot; &quot;A&quot; &quot;G&quot; ...
##  $ source                      : chr  &quot;cosmic&quot; &quot;cosmic&quot; &quot;cosmic&quot; &quot;cosmic&quot; ...
##  $ geneName                    : chr  &quot;TET1&quot; &quot;TET1&quot; &quot;TET1&quot; &quot;TET1&quot; ...
##  $ mutationGRCh37Strand        : chr  &quot;+&quot; &quot;+&quot; &quot;+&quot; &quot;+&quot; ...
##  $ primarySite                 : chr  &quot;lung&quot; &quot;ovary&quot; &quot;ovary&quot; &quot;prostate&quot; ...
##  $ mutationAA                  : chr  &quot;p.D1715N&quot; &quot;p.H1706Q&quot; &quot;p.A2023T&quot; &quot;p.F329V&quot; ...
##  $ tumourOrigin                : chr  &quot;primary&quot; &quot;primary&quot; &quot;primary&quot; &quot;primary&quot; ...
##  $ histologySubtype            : chr  &quot;adenocarcinoma&quot; &quot;serous_carcinoma&quot; &quot;serous_carcinoma&quot; &quot;NS&quot; ...
##  $ accessionNumber             : chr  &quot;ENST00000373644&quot; &quot;ENST00000373644&quot; &quot;ENST00000373644&quot; &quot;ENST00000373644&quot; ...
##  $ mutationID                  : chr  &quot;COSM94115&quot; &quot;COSM76709&quot; &quot;COSM72946&quot; &quot;COSM94112&quot; ...
##  $ mutationCDS                 : chr  &quot;c.5143G&gt;A&quot; &quot;c.5118T&gt;A&quot; &quot;c.6067G&gt;A&quot; &quot;c.985T&gt;G&quot; ...
##  $ sampleName                  : chr  &quot;94699&quot; &quot;TCGA-24-0966-01&quot; &quot;TCGA-13-0899-01&quot; &quot;111995&quot; ...
##  $ primaryHistology            : chr  &quot;carcinoma&quot; &quot;carcinoma&quot; &quot;carcinoma&quot; &quot;carcinoma&quot; ...
##  $ mutationGRCh37GenomePosition: chr  &quot;10:70446203-70446203&quot; &quot;10:70446178-70446178&quot; &quot;10:70451227-70451227&quot; &quot;10:70333081-70333081&quot; ...
##  $ mutationDescription         : chr  &quot;Substitution - Missense&quot; &quot;Substitution - Missense&quot; &quot;Substitution - Missense&quot; &quot;Substitution - Missense&quot; ...
##  $ genomeWideScreen            : chr  &quot;n&quot; &quot;y&quot; &quot;y&quot; &quot;n&quot; ...
##  $ idTumour                    : chr  &quot;1443016&quot; &quot;1398672&quot; &quot;1398733&quot; &quot;1443190&quot; ...
##  $ idSample                    : chr  &quot;1520641&quot; &quot;1474973&quot; &quot;1475034&quot; &quot;1520815&quot; ...
##  $ mutationSomaticStatus       : chr  &quot;Confirmed somatic variant&quot; &quot;Confirmed somatic variant&quot; &quot;Confirmed somatic variant&quot; &quot;Confirmed somatic variant&quot; ...
##  $ siteSubtype                 : chr  &quot;NS&quot; &quot;NS&quot; &quot;NS&quot; &quot;NS&quot; ...
##  $ geneCDSLength               : int  6411 6411 6411 6411 6411 6411 6411 6411 6411 6411 ...
##  $ hgncId                      : chr  &quot;29484&quot; &quot;29484&quot; &quot;29484&quot; &quot;29484&quot; ...
##  $ pubmedPMID                  : chr  &quot;20668451&quot; &quot;21720365&quot; &quot;21720365&quot; &quot;20668451&quot; ...
##  $ sampleSource                : chr  &quot;NS&quot; &quot;NS&quot; &quot;NS&quot; &quot;NS&quot; ...
##  $ snp                         : logi  FALSE FALSE FALSE FALSE FALSE FALSE ...
##  $ fathmmPrediction            : chr  &quot;PATHOGENIC&quot; &quot;NEUTRAL&quot; &quot;PATHOGENIC&quot; &quot;NEUTRAL&quot; ...
##  $ _geneIds                    :List of 12
##  $ _phenotypes                 :List of 12
##  $ mutationZygosity            : chr  NA &quot;het&quot; &quot;het&quot; NA ...
##  $ age                         : num  NA 78 60 NA 73 NA NA 69 NA 39 ...
##  $ idStudy                     : int  NA 331 331 NA 331 NA NA 331 NA 331 ...</code></pre>
</div>
</div>
</div>
<div id="cellbaser-utilities" class="section level1">
<h1><span class="header-section-number">3</span> CellbaseR utilities</h1>
<div id="creategenemodel" class="section level2">
<h2><span class="header-section-number">3.1</span> CreateGeneModel</h2>
<p>A usefull utility for fast building of gene models, compatible with Gviz package for genomic visualization</p>
<pre class="r"><code>library(cellbaseR)
cb &lt;- CellBaseR()
test &lt;- createGeneModel(object = cb, region = &quot;17:1500000-1550000&quot;)
library(Gviz)
testTrack &lt;- GeneRegionTrack(test)
plotTracks(testTrack, transcriptAnnotation='symbol')
# then you can use this track in conjunction with any other tracks
ideoTrack &lt;- IdeogramTrack(genome = &quot;hg19&quot;, chromosome = &quot;chr17&quot;)
axisTrack &lt;- GenomeAxisTrack()
from &lt;- min(test$start)-5000
to &lt;- max(test$end)+5000
plotTracks(list(ideoTrack,axisTrack,testTrack),from = from, to = to, transcriptAnnotation='symbol')</code></pre>
</div>
<div id="cbannotatevcf" class="section level2">
<h2><span class="header-section-number">3.2</span> cbAnnotateVcf</h2>
<p>This utility allows users to annoate genomic variants from small to medium-sized vcf files directly from the cellbase server, with a wealth of genomic annotations including riche clinical annotations like clinvar, gwas, and cosmic data, as well as conservation and functional scores like phast and cadd scores, without the need to download any files.</p>
<pre class="r"><code>library(cellbaseR)
cb &lt;- CellBaseR()
fl &lt;- system.file(&quot;extdata&quot;, &quot;chr22.vcf.gz&quot;, package=&quot;VariantAnnotation&quot;)
res &lt;- cbAnnotateVcf(object=cb, file=fl)
res &lt;- cbData(res)
res</code></pre>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
