\nonstopmode{}
\documentclass[letterpaper]{book}
\usepackage[times,inconsolata,hyper]{Rd}
\usepackage{makeidx}
\usepackage[utf8,latin1]{inputenc}
% \usepackage{graphicx} % @USE GRAPHICX@
\makeindex{}
\begin{document}
\chapter*{}
\begin{center}
{\textbf{\huge Package `cellbaseR'}}
\par\bigskip{\large \today}
\end{center}
\begin{description}
\raggedright{}
\item[Type]\AsIs{Package}
\item[Date]\AsIs{2016-04-17}
\item[Title]\AsIs{Querying annotation data from the high performance Cellbase web
services}
\item[Version]\AsIs{1.1.0}
\item[Author]\AsIs{Mohammed OE Abdallah}
\item[Maintainer]\AsIs{Mohammed OE Abdallah }\email{melsiddieg@gmail.com}\AsIs{}
\item[URL]\AsIs{}\url{https://github.com/melsiddieg/cellbaseR}\AsIs{}
\item[Description]\AsIs{This R package makes use of the exhaustive RESTful Web
service API that has been implemented for the Cellabase
database. It enable researchers to query and obtain a wealth of
biological information from a single database saving a lot of
time. Another benefit is that researchers can easily make
queries about different biological topics and link all this
information together as all information is integrated.}
\item[License]\AsIs{Apache License (== 2.0)}
\item[Depends]\AsIs{R(>= 3.4)}
\item[Imports]\AsIs{methods, jsonlite, httr, data.table, pbapply, tidyr, R.utils,
Rsamtools, BiocParallel, foreach, utils, parallel, doParallel}
\item[Suggests]\AsIs{BiocStyle, knitr, rmarkdown, Gviz, VariantAnnotation}
\item[RoxygenNote]\AsIs{6.0.1}
\item[biocViews]\AsIs{Annotation, VariantAnnotation}
\item[Lazy]\AsIs{TRUE}
\item[LazyData]\AsIs{TRUE}
\item[Collate]\AsIs{'commons.R' 'AllClasses.R' 'AllGenerics.R'
'AnnotateVcf-methods.R' 'CellBaseParam-methods.R'
'CellBaseR-methods.R' 'cellbase.R' 'getCellbase-methods.R'
'getChromosomeInfo-methods.R' 'getClinical-methods.R'
'getGene-methods.R' 'getMeta-methods.R' 'getProtein-methods.R'
'getRegion-methods.R' 'getSnp-methods.R' 'getTf-methods.R'
'getTranscript-methods.R' 'getVariant-methods.R'
'getXref-methods.R' 'show-methods.R' 'tools.R' 'user.R'}
\item[VignetteBuilder]\AsIs{knitr}
\item[NeedsCompilation]\AsIs{no}
\end{description}
\Rdcontents{\R{} topics documented:}
\inputencoding{utf8}
\HeaderA{cellbaseR-package}{cellbaseR}{cellbaseR.Rdash.package}
\aliasA{cellbaseR}{cellbaseR-package}{cellbaseR}
%
\begin{Description}\relax
Querying annotation data from the high performance Cellbase web
services
\end{Description}
%
\begin{Details}\relax
Documentation for the cellbaseR package

This R package makes use of the exhaustive RESTful Web service 
API that has been
implemented for the Cellabase database. It enables researchers to query and
obtain a wealth of biological information from a single database saving a lot 
of time. Another benefit is that researchers can easily make  queries about 
different biological topics and link all this information together as all 
information is integrated.
Currently Homo sapiens, Mus musculus and other 20 species are available and 
many others will be included soon. Results returned from the cellbase queries
are parsed into R data.frames and other common R data strctures so users can 
readily get into downstream anaysis.
\end{Details}
%
\begin{Author}\relax
Mohammed OE Abdallah
\end{Author}
%
\begin{SeeAlso}\relax
Useful links:
\begin{itemize}

\item \url{https://github.com/melsiddieg/cellbaseR}

\end{itemize}


\end{SeeAlso}
\inputencoding{utf8}
\HeaderA{AnnotateVcf,CellBaseR-method}{AnnotateVcf}{AnnotateVcf,CellBaseR.Rdash.method}
\aliasA{AnnotateVcf}{AnnotateVcf,CellBaseR-method}{AnnotateVcf}
%
\begin{Description}\relax
This method is a convience method to annotate bgzipped tabix-indexed vcf 
files. It should be ideal for annotating small to medium sized 
vcf files.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
## S4 method for signature 'CellBaseR'
AnnotateVcf(object, file, batch_size, num_threads,
  BPPARAM = bpparam())
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{file}] Path to a bgzipped and tabix indexed vcf file

\item[\code{batch\_size}] intger if multiple queries are raised by a single method 
call, e.g. getting annotation info for several genes,
queries will be sent to the server in batches. This slot indicates the size 
of each batch, e.g. 200

\item[\code{num\_threads}] number of asynchronus batches to be sent to the server

\item[\code{BPPARAM}] a BiocParallel class object
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a dataframe with the results of the query
\end{Value}
%
\begin{SeeAlso}\relax
\url{https://github.com/opencb/cellbase/wiki} 
and the RESTful API documentation 
\url{http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
cb <- CellBaseR()
fl <- system.file("extdata", "hapmap_exome_chr22_500.vcf.gz",
                  package = "cellbaseR" )
res <- AnnotateVcf(object=cb, file=fl, BPPARAM = bpparam(workers=2))
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{CellBaseParam}{A Constructor for the CellBaseParam Object}{CellBaseParam}
%
\begin{Description}\relax
CellBaseParam object is used to control what results are returned from the
CellBaseR methods
\end{Description}
%
\begin{Usage}
\begin{verbatim}
CellBaseParam(genome = character(), gene = character(),
  region = character(), rs = character(), so = character(),
  phenotype = character(), include = character(), exclude = character(),
  limit = character())
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{genome}] A character denoting the genome build to query,eg, GRCh37
(default),or GRCh38

\item[\code{gene}] A character vector denoting the gene/s to be queried

\item[\code{region}] A character vector denoting the region/s to be queried must be
in the form 1:100000-1500000 not chr1:100000-1500000

\item[\code{rs}] A character vector denoting the rs ids to be queried

\item[\code{so}] A character vector denoting sequence ontology to be queried

\item[\code{phenotype}] A character vector denoting the phenotype to be queried

\item[\code{include}] A character vector denoting the fields to be returned

\item[\code{exclude}] A character vector denoting the fields to be excluded

\item[\code{limit}] A number limiting the number of results to be returned
\end{ldescription}
\end{Arguments}
%
\begin{Value}
an object of class CellBaseParam
\end{Value}
%
\begin{SeeAlso}\relax
\url{https://github.com/opencb/cellbase/wiki} 
and the RESTful API documentation 
\url{http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
cbParam <- CellBaseParam(genome="GRCh38",gene=c("TP73","TET1"))
print(cbParam)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{CellBaseParam-class}{CellBaseParam Class}{CellBaseParam.Rdash.class}
%
\begin{Description}\relax
This class  defines a CellBaseParam object to hold filtering 
parameters.
\end{Description}
%
\begin{Details}\relax
This class stores parameters used for filtering the CellBaseR query
and is avaialable for all query methods. CellBaseParam object is used to
control what results are returned from the' CellBaseR methods
\end{Details}
%
\begin{Section}{Slots}

\begin{description}

\item[\code{genome}] A character the genome build to query, e.g.GRCh37(default)

\item[\code{gene}] A character vector denoting the gene/s to be queried

\item[\code{region}] A character vector denoting the region/s to be queried must be 
in the form 1:100000-1500000

\item[\code{rs}] A character vector denoting the rs ids to be queried

\item[\code{so}] A character vector denoting sequence ontology to be queried

\item[\code{phenotype}] A character vector denoting the phenotype to be queried

\item[\code{include}] A character vector denoting the fields to be returned

\item[\code{exclude}] A character vector denoting the fields to be excluded

\item[\code{limit}] A number limiting the number of results to be returned

\end{description}
\end{Section}
%
\begin{SeeAlso}\relax
\url{https://github.com/opencb/cellbase/wiki} 
and the RESTful API documentation 
\url{http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/}
\end{SeeAlso}
\inputencoding{utf8}
\HeaderA{CellBaseR}{CellBaseR}{CellBaseR}
%
\begin{Description}\relax
This is a constructor function for the CellBaseR object
\end{Description}
%
\begin{Usage}
\begin{verbatim}
CellBaseR(host = "http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/rest/",
  version = "v4", species = "hsapiens", batch_size = 200L,
  num_threads = 8L)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{host}] A character the default host url for cellbase webservices,
e.g. "http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/rest/"

\item[\code{version}] A character the cellbae API version, e.g. "V4"

\item[\code{species}] a character specifying the species to be queried, e.g. 
"hsapiens"

\item[\code{batch\_size}] intger if multiple queries are raised by a single method 
call, e.g. getting annotation info for several genes, queries will be sent 
to the server in batches.This slot indicates the size of each batch,e.g. 200

\item[\code{num\_threads}] integer number of  batches to be sent to the server
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
CellbaseR constructor function


This class defines the CellBaseR object. It holds the default
configuration required by CellBaseR methods to connect to the
cellbase web services. By defult it is configured to query human
data based on the GRCh37 genome assembly.
\end{Details}
%
\begin{Value}
An object of class CellBaseR
\end{Value}
%
\begin{SeeAlso}\relax
\url{https://github.com/opencb/cellbase/wiki} 
and the RESTful API documentation 
\url{http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
   cb <- CellBaseR()
   print(cb)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{CellBaseR-class}{CellBaseR Class}{CellBaseR.Rdash.class}
%
\begin{Description}\relax
This is an S4 class  which defines the CellBaseR object
\end{Description}
%
\begin{Details}\relax
This S4 class holds the default configuration required by CellBaseR 
methods to connect to the cellbase web 
services. By default it is configured to query human data based on the GRCh37
genome assembly.
\end{Details}
%
\begin{Section}{Slots}

\begin{description}

\item[\code{host}] a character specifying the host url. Default 
"http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/rest/"

\item[\code{version}] a character specifying the API version. Default "v4"

\item[\code{species}] a character specifying the species to be queried. Default
"hsapiens"

\item[\code{batch\_size}] if multiple queries are raised by a single method call, e.g.
getting annotation info for several genes,
queries will be sent to the server in batches. This slot indicates the size 
of these batches. Default 200

\item[\code{num\_threads}] the number of threads. Default 8

\end{description}
\end{Section}
%
\begin{SeeAlso}\relax
\url{https://github.com/opencb/cellbase/wiki} 
and the RESTful API documentation 
\url{http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/}
\end{SeeAlso}
\inputencoding{utf8}
\HeaderA{createGeneModel}{createGeneModel}{createGeneModel}
%
\begin{Description}\relax
A convience functon to construct a genemodel
\end{Description}
%
\begin{Usage}
\begin{verbatim}
createGeneModel(object, region = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellbaseResponse

\item[\code{region}] a character
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function create a gene model data frame, which can be then 
turned into a GeneRegionTrack for visualiaztion
by \code{\LinkA{GeneRegionTrack}{GeneRegionTrack}}
\end{Details}
%
\begin{Value}
A geneModel
\end{Value}
%
\begin{SeeAlso}\relax
\url{https://github.com/opencb/cellbase/wiki} 
and the RESTful API documentation 
\url{http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
cb <- CellBaseR()
test <- createGeneModel(object = cb, region = "17:1500000-1550000")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getCaddScores}{getCaddScores}{getCaddScores}
%
\begin{Description}\relax
A convienice method to fetch Cadd scores for specific variant/s
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getCaddScores(object, id, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{id}] a charcter vector of genomic variants, eg 19:45411941:T:C

\item[\code{param}] an object of class CellBaseParam
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a dataframe of the query result
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
cb <- CellBaseR()
res <- getCaddScores(cb, "19:45411941:T:C")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getCellBase,CellBaseR-method}{getCellBase}{getCellBase,CellBaseR.Rdash.method}
\aliasA{getCellBase}{getCellBase,CellBaseR-method}{getCellBase}
%
\begin{Description}\relax
The generic method for querying CellBase web services.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
## S4 method for signature 'CellBaseR'
getCellBase(object, category, subcategory, ids, resource,
  param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{category}] character to specify the category to be queried.

\item[\code{subcategory}] character to specify the subcategory to be queried

\item[\code{ids}] a character vector of the ids to be queried

\item[\code{resource}] a character to specify the resource to be queried

\item[\code{param}] an object of class CellBaseParam specifying additional param
for the CellBaseR
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This method allows the user to query the 
cellbase web services without any  predefined categories, subcategries, 
or resources.
\end{Details}
%
\begin{Value}
a dataframe holding the results of the query
\end{Value}
%
\begin{SeeAlso}\relax
\url{https://github.com/opencb/cellbase/wiki} 
and the RESTful API documentation 
\url{http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
   cb <- CellBaseR()
   res <- getCellBase(object=cb, category="feature", subcategory="gene", 
   ids="TET1", resource="info")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getCellBaseResourceHelp}{getCellBaseResourceHelp}{getCellBaseResourceHelp}
%
\begin{Description}\relax
A function to get help about available cellbase resources
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getCellBaseResourceHelp(object, subcategory)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] a cellBase class object

\item[\code{subcategory}] a character the subcategory to be queried
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This function retrieves available resources for each generic method
like getGene, getRegion, getprotein, etc. It help the user see all possible 
resources to use with the getGeneric methods
\end{Details}
%
\begin{Value}
character vector of the available resources to that particular 
subcategory
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
cb <- CellBaseR()
# Get help about what resources are available to the getGene method
getCellBaseResourceHelp(cb, subcategory="gene")
# Get help about what resources are available to the getRegion method
getCellBaseResourceHelp(cb, subcategory="region")
# Get help about what resources are available to the getXref method
getCellBaseResourceHelp(cb, subcategory="id")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getChromosomeInfo,CellBaseR-method}{getChromosomeInfo}{getChromosomeInfo,CellBaseR.Rdash.method}
\aliasA{getChromosomeInfo}{getChromosomeInfo,CellBaseR-method}{getChromosomeInfo}
%
\begin{Description}\relax
A method to query sequence data from Cellbase web services.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
## S4 method for signature 'CellBaseR'
getChromosomeInfo(object, ids, resource, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{ids}] a character vector of chromosome ids to be queried

\item[\code{resource}] a character vector to specify the resource to be queried

\item[\code{param}] a object of class CellBaseParam specifying additional param for
the query
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
A method to query sequence data from Cellbase web services. This 
method retrieves information about chromosomes, including its size and
detailed information about its different cytobands
\end{Details}
%
\begin{Value}
a dataframe with the results of the query
\end{Value}
%
\begin{SeeAlso}\relax
\url{https://github.com/opencb/cellbase/wiki} 
and the RESTful API documentation 
\url{http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
   cb <- CellBaseR()
   res <- getChromosomeInfo(object=cb, ids="22", resource="info")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getClinical,CellBaseR-method}{getClinical}{getClinical,CellBaseR.Rdash.method}
\aliasA{getClinical}{getClinical,CellBaseR-method}{getClinical}
%
\begin{Description}\relax
A method to query Clinical data from Cellbase web services.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
## S4 method for signature 'CellBaseR'
getClinical(object, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{param}] a object of class CellBaseParam specifying the parameters
limiting the CellBaseR
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This method retrieves clinicaly relevant variants annotations from
multiple resources including clinvar, cosmic and gwas catalog. Furthermore,
the user can filter these data in many ways including phenotype, genes, rs,
etc,.
\end{Details}
%
\begin{Value}
a dataframe with the results of the query
\end{Value}
%
\begin{SeeAlso}\relax
\url{https://github.com/opencb/cellbase/wiki} 
and the RESTful API documentation 
\url{http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
   cb <- CellBaseR()
   cbParam <- CellBaseParam(gene=c("TP73","TET1"))
   res <- getClinical(object=cb,param=cbParam)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getClinicalByGene}{getClinicalByGene}{getClinicalByGene}
%
\begin{Description}\relax
A convienice method to fetch clinical variants for specific gene/s
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getClinicalByGene(object, id, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of CellBaseR class

\item[\code{id}] a charcter vector of HUGO symbol (gene names)

\item[\code{param}] an object of class CellBaseParam
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a dataframe of the query result
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
cb <- CellBaseR()
res <- getClinicalByGene(cb, "TET1")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getClinicalByRegion}{getClinicalByRegion}{getClinicalByRegion}
%
\begin{Description}\relax
A convienice method to fetch clinical variants for specific region/s
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getClinicalByRegion(object, id, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{id}] a charcter vector of genomic regions, eg 17:1000000-1100000

\item[\code{param}] an object of class CellBaseParam
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a dataframe of the query result
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
cb <- CellBaseR()
res <- getClinicalByRegion(cb, "17:1000000-1189811")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getConservationByRegion}{getConservationByRegion}{getConservationByRegion}
%
\begin{Description}\relax
A convienice method to fetch conservation data for specific region/s
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getConservationByRegion(object, id, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{id}] a charcter vector of genomic regions, eg 17:1000000-1100000

\item[\code{param}] an object of class CellBaseParam
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a dataframe of the query result
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
cb <- CellBaseR()
res <- getConservationByRegion(cb, "17:1000000-1189811")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getGene,CellBaseR-method}{getGene}{getGene,CellBaseR.Rdash.method}
\aliasA{getGene}{getGene,CellBaseR-method}{getGene}
%
\begin{Description}\relax
A method to query gene data from Cellbase web services.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
## S4 method for signature 'CellBaseR'
getGene(object, ids, resource, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{ids}] a character vector of gene ids to be queried

\item[\code{resource}] a character vector to specify the resource to be queried

\item[\code{param}] an object of class CellBaseParam specifying additional param
for the CellBaseR
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This method retrieves various gene annotations including transcripts
and exons data as well as gene expression and clinical data
\end{Details}
%
\begin{Value}
a dataframe with the results of the query
\end{Value}
%
\begin{SeeAlso}\relax
\url{https://github.com/opencb/cellbase/wiki} 
and the RESTful API documentation 
\url{http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
   cb <- CellBaseR()
   res <- getGene(object=cb, ids=c("TP73","TET1"), resource="info")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getGeneInfo}{getGeneInfo}{getGeneInfo}
%
\begin{Description}\relax
A convienice method to fetch gene annotations specific gene/s
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getGeneInfo(object, id, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{id}] a charcter vector of HUGO symbol (gene names)

\item[\code{param}] an object of class CellBaseParam
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a dataframe of the query result
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
cb <- CellBaseR()
res <- getGeneInfo(cb, "TET1")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getMeta,CellBaseR-method}{getMeta}{getMeta,CellBaseR.Rdash.method}
\aliasA{getMeta}{getMeta,CellBaseR-method}{getMeta}
%
\begin{Description}\relax
A method for getting the available metadata from the cellbase web services
\end{Description}
%
\begin{Usage}
\begin{verbatim}
## S4 method for signature 'CellBaseR'
getMeta(object, resource)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{resource}] the resource you want to query it metadata
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This method is for getting information about the avaialable species
and available annotation, assembly for each species from the cellbase web 
services.
\end{Details}
%
\begin{Value}
a dataframe with the
results of the query
\end{Value}
%
\begin{SeeAlso}\relax
\url{https://github.com/opencb/cellbase/wiki} 
and the RESTful API documentation 
\url{http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
   cb <- CellBaseR()
   res <- getMeta(object=cb, resource="species")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getProtein,CellBaseR-method}{getProtein}{getProtein,CellBaseR.Rdash.method}
\aliasA{getProtein}{getProtein,CellBaseR-method}{getProtein}
%
\begin{Description}\relax
A method to query protein data from Cellbase web services.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
## S4 method for signature 'CellBaseR'
getProtein(object, ids, resource, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{ids}] a character vector of uniprot ids to be queried, should be one
or more of uniprot ids, for example O15350.

\item[\code{resource}] a character vector to specify the resource to be queried

\item[\code{param}] a object of class CellBaseParam specifying additional param
for the query
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This method retrieves various protein annotations including 
protein description, features, sequence, substitution scores, evidence,  etc.
\end{Details}
%
\begin{Value}
an object of class CellBaseResponse which holds a dataframe with th
e results of the query
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
   cb <- CellBaseR()
   res <- getProtein(object=cb, ids="O15350", resource="info")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getProteinInfo}{getProteinInfo}{getProteinInfo}
%
\begin{Description}\relax
A convienice method to fetch annotations for specific protein/s
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getProteinInfo(object, id, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{id}] a charcter vector of Uniprot Ids

\item[\code{param}] an object of class CellBaseParam
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a dataframe of the query result
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
cb <- CellBaseR()
res <- getProteinInfo(cb, "O15350")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getRegion,CellBaseR-method}{getRegion}{getRegion,CellBaseR.Rdash.method}
\aliasA{getRegion}{getRegion,CellBaseR-method}{getRegion}
%
\begin{Description}\relax
A method to query features within a genomic region from Cellbase web 
services.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
## S4 method for signature 'CellBaseR'
getRegion(object, ids, resource, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{ids}] a character vector of the regions to be queried, for example,
"1:1000000-1200000' should always be in the
form 'chr:start-end'

\item[\code{resource}] a character vector to specify the resource to be queried

\item[\code{param}] a object of class CellBaseParam specifying additional param
for the query
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This method retrieves various genomic features from a given
region including genes, snps, clincally relevant variants, proteins, etc.
\end{Details}
%
\begin{Value}
a dataframe with the results of the query
\end{Value}
%
\begin{SeeAlso}\relax
\url{https://github.com/opencb/cellbase/wiki} 
and the RESTful API documentation 
\url{http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
   cb <- CellBaseR()
   res <- getRegion(object=cb, ids="17:1000000-1200000", resource="gene")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getRegulatoryByRegion}{getRegulatoryByRegion}{getRegulatoryByRegion}
%
\begin{Description}\relax
A convienice method to fetch regulatory data for specific region/s
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getRegulatoryByRegion(object, id, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{id}] a charcter vector of genomic regions, eg 17:1000000-1100000

\item[\code{param}] an object of class CellBaseParam
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a dataframe of the query result
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
cb <- CellBaseR()
res <- getRegulatoryByRegion(cb, "17:1000000-1189811")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getSnp,CellBaseR-method}{getSnp}{getSnp,CellBaseR.Rdash.method}
\aliasA{getSnp}{getSnp,CellBaseR-method}{getSnp}
%
\begin{Description}\relax
A method to query genomic variation data from Cellbase web services.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
## S4 method for signature 'CellBaseR'
getSnp(object, ids, resource, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{ids}] a character vector of the ids to be queried, must be a valid rsid,
for example 'rs6025'

\item[\code{resource}] a character vector to specify the resource to be queried

\item[\code{param}] a object of class CellBaseParam specifying additional param
for the query
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
.

This method retrieves known genomic variants (snps) and their
annotations including population frequncies from 1k genomes and Exac projects
as well as clinical data and various other annotations
\end{Details}
%
\begin{Value}
a dataframe with the results of the query
\end{Value}
%
\begin{SeeAlso}\relax
\url{https://github.com/opencb/cellbase/wiki} 
and the RESTful API documentation 
\url{http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
cb <- CellBaseR()
res <- getSnp(object=cb, ids="rs6025", resource="info")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getSnpByGene}{getSnpByGene}{getSnpByGene}
%
\begin{Description}\relax
A convienice method to fetch known variants (snps) for specific gene/s
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getSnpByGene(object, id, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{id}] a charcter vector of HUGO symbol (gene names)

\item[\code{param}] an object of class CellBaseParam
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a dataframe of the query result
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
cb <- CellBaseR()
param <- CellBaseParam(limit = 10)
res <- getSnpByGene(cb, "TET1", param = param)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getTf,CellBaseR-method}{getTf}{getTf,CellBaseR.Rdash.method}
\aliasA{getTf}{getTf,CellBaseR-method}{getTf}
%
\begin{Description}\relax
A method to query transcription factors binding sites data from Cellbase web
services.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
## S4 method for signature 'CellBaseR'
getTf(object, ids, resource, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{ids}] a character vector of the ids to be queried, must be a valid 
transcription factor name, for example, 
eg, CTCF

\item[\code{resource}] a character vector to specify the resource to be queried

\item[\code{param}] a object of class CellBaseParam specifying additional param
for the query
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This method retrieves various transcription factors binding sites 
data
\end{Details}
%
\begin{Value}
a dataframe with the results of the query
\end{Value}
%
\begin{SeeAlso}\relax
\url{https://github.com/opencb/cellbase/wiki} 
and the RESTful API documentation 
\url{http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
   cb <- CellBaseR()
   param <- CellBaseParam(limit = 12)
   res <- getTf(object=cb, ids="CTCF", resource="tfbs", param=param)
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getTfbsByRegion}{getTfbsByRegion}{getTfbsByRegion}
%
\begin{Description}\relax
A convienice method to fetch Transcription facrots data for specific region/s
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getTfbsByRegion(object, id, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{id}] a charcter vector of genomic regions, eg 17:1000000-1100000

\item[\code{param}] an object of class CellBaseParam
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a dataframe of the query result
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
cb <- CellBaseR()
res <- getTfbsByRegion(cb, "17:1000000-1189811")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getTranscript,CellBaseR-method}{getTranscript}{getTranscript,CellBaseR.Rdash.method}
\aliasA{getTranscript}{getTranscript,CellBaseR-method}{getTranscript}
%
\begin{Description}\relax
A method to query transcript data from Cellbase web services.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
## S4 method for signature 'CellBaseR'
getTranscript(object, ids, resource, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{ids}] a character vector of the transcript ids to be queried, use 
ensemble transccript IDs eq, ENST00000380152

\item[\code{resource}] a character vector to specify the resource to be queried

\item[\code{param}] an object of class CellBaseParam specifying additional params 
for the query
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This method retrieves various genomic annotations for transcripts
including exons, cDNA sequence, annotations flags, and cross references,etc.
\end{Details}
%
\begin{Value}
a dataframe with the results of the query
\end{Value}
%
\begin{SeeAlso}\relax
\url{https://github.com/opencb/cellbase/wiki} 
and the RESTful API documentation 
\url{http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
   cb <- CellBaseR()
   res <- getTranscript(object=cb, ids="ENST00000373644", resource="info")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getTranscriptByGene}{getTranscriptByGene}{getTranscriptByGene}
%
\begin{Description}\relax
A convienice method to fetch transcripts for specific gene/s
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getTranscriptByGene(object, id, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{id}] a charcter vector of HUGO symbol (gene names)

\item[\code{param}] an object of class CellBaseParam
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a dataframe of the query result
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
cb <- CellBaseR()
res <- getTranscriptByGene(cb, "TET1")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getVariant,CellBaseR-method}{getVariant}{getVariant,CellBaseR.Rdash.method}
\aliasA{getVariant}{getVariant,CellBaseR-method}{getVariant}
%
\begin{Description}\relax
A method to query variant annotation data from Cellbase web services from
Cellbase web services.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
## S4 method for signature 'CellBaseR'
getVariant(object, ids, resource, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{ids}] a character vector of the ids to be queried, must be in the 
following format 'chr:start:ref:alt', for 
example, '1:128546:A:T'

\item[\code{resource}] a character vector to specify the resource to be queried

\item[\code{param}] a object of class CellBaseParam specifying additional param
for the query
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This method retrieves extensive genomic annotations for variants
including consequence types, conservation data, population frequncies from 1k
genomes and Exac projects, etc.
as well as clinical data and various other annotations
\end{Details}
%
\begin{Value}
a dataframe with the results of the query
\end{Value}
%
\begin{SeeAlso}\relax
\url{https://github.com/opencb/cellbase/wiki} 
and the RESTful API documentation 
\url{http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
   cb <- CellBaseR()
   res <- getVariant(object=cb, ids="19:45411941:T:C", resource="annotation")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getVariantAnnotation}{getVariantAnnotation}{getVariantAnnotation}
%
\begin{Description}\relax
A convienice method to fetch variant annotation for specific variant/s
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getVariantAnnotation(object, id, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{id}] a charcter vector of length < 200 of genomic variants,
eg 19:45411941:T:C

\item[\code{param}] an object of class CellBaseParam
\end{ldescription}
\end{Arguments}
%
\begin{Value}
a dataframe of the query result
\end{Value}
%
\begin{Examples}
\begin{ExampleCode}
cb <- CellBaseR()
res <- getVariantAnnotation(cb, "19:45411941:T:C")
\end{ExampleCode}
\end{Examples}
\inputencoding{utf8}
\HeaderA{getXref,CellBaseR-method}{getXref}{getXref,CellBaseR.Rdash.method}
\aliasA{getXref}{getXref,CellBaseR-method}{getXref}
%
\begin{Description}\relax
A method to query cross reference data from Cellbase web services.
\end{Description}
%
\begin{Usage}
\begin{verbatim}
## S4 method for signature 'CellBaseR'
getXref(object, ids, resource, param = NULL)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{object}] an object of class CellBaseR

\item[\code{ids}] a character vector of the ids to be queried, any crossrefereable
ID, gene names, transcript ids, 
uniprot ids,etc.

\item[\code{resource}] a character vector to specify the resource to be queried

\item[\code{param}] a object of class CellBaseParam specifying additional param
for the query
\end{ldescription}
\end{Arguments}
%
\begin{Details}\relax
This method retrieves cross references for genomic identifiers, eg
ENSEMBL ids, it also provide starts\_with service that is useful for
autocomplete services.
\end{Details}
%
\begin{Value}
a dataframe with the results of the query
\end{Value}
%
\begin{SeeAlso}\relax
\url{https://github.com/opencb/cellbase/wiki} 
and the RESTful API documentation 
\url{http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/}
\end{SeeAlso}
%
\begin{Examples}
\begin{ExampleCode}
   cb <- CellBaseR()
   res <- getXref(object=cb, ids="ENST00000373644", resource="xref")
\end{ExampleCode}
\end{Examples}
\printindex{}
\end{document}
